Start SecureNetworkInterface

IgorFilipovich opened the pull request:
Add ecnryption with use of AES256 symmetric ecnryption and DiffieHellman
assymetric encryption.
Implement Dismess Protocol.
Start SessionManager

IgorFilipovich committed:
Start SecurityNetworkInterface

Add ecnryption with use of AES256 symmetric ecnryption and DiffieHellman
assymetric encryption.
Implement Dismess Protocol.
Start SessionManager

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,90 @@
+        return (0..rightBorder).random()
+    }
+
+    fun getPublicKey(): ByteArray {
```
Это Java-style getter, нужно сделать publicKey публичным property и переопределить у него getter, а setter сделать правитным (private set)

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    fun encrypt(message: ByteArray): ByteArray {
+        println(aesSendOffset)
```
Можешь подключить logger в отдельной ветке из develop, а потом слить develop сюда. https://www.baeldung.com/kotlin/logging

Derovi commented on src/main/kotlin/by/dismess/core/security/ProtocolManager.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    fun processData(data: ByteArray): ByteArray? {
+        if (data[0].compareTo(1) == 0) {
```
data.first() == 0.toByte()
В котлине переопределен оператор == для всех многих типов, тем более для примитивов. А скомпилировать не получалось потому что в котлине нет имплисит-кастов, т.е. нужно явно указать 0.toByte().

Derovi commented on src/main/kotlin/by/dismess/core/security/SessionManager.kt:
```diff
@@ -0,0 +1,33 @@
+        return protocolManager.updateKey()
+    }
+
+    fun encrypt(address: InetSocketAddress, data: ByteArray): ByteArray {
```
1) Если ты уверен, что всегда в мапе есть нужный ключ, то лучше используй .getValue(address)
2) Кажется, всю функцию можно в одну строку записать. просто = ...getValue(address).encrypt(data)

Derovi reviewed (commented):
Поверхностно прочитал, лучше ревьювить в конце рабочую версию.
Пожалуйста, постарайся добавить как можно больше комментов, код шифрования сложно без них читать =)
Добавь юнит-тест, это сэкономит много времени при тестировании. (см. как сделаны тесты в пакете managers)

IgorFilipovich committed:
Secure Network Interface continuation

Change DISMESS protocol
Change session managment system

IgorFilipovich committed:
Finish SecureNetworkInterface

Add unit-tests for Encryptor and SecureNetworkInterface
Finish SecureNetworkInterface with session management

IgorFilipovich committed:
Fix ktlint

buyval01 commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,142 @@
+        return SecretKeySpec(aesKey, aesOffset, AES_SIZE, cipherAlgo)
+    }
+
+    private fun findSession(sessionNumber: Int): ByteArray? {
```
По-моему, тут ищется ключ конкретной сессии, поэтому предлогаю переименовать в  findSessionKey

VKuzia replied:
одобряю

buyval01 commented on src/main/kotlin/by/dismess/core/security/SecureNetworkInterface.kt:
```diff
@@ -0,0 +1,72 @@
+
+    private fun inetAddressToInetSocketAddress(address: InetAddress): InetSocketAddress {
+        return InetSocketAddress(1)
+    }
```
Это вообще для чего?)

IgorFilipovich replied:
Это заглушка для тестов, когда Рома сольет пофиксится

buyval01 commented on src/test/kotlin/by/dismess/core/security/SecureNetworkInterfaceTest.kt:
```diff
@@ -0,0 +1,76 @@
+        firstCounter: Int
+    ) {
+        val firstAddress = InetSocketAddress(1)
+        val message = "Дарова".toByteArray()
```
Привет

IgorFilipovich replied:
А я тебя знаю

VKuzia commented on src/test/kotlin/by/dismess/core/security/EncryptorTest.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Test
+    fun testSimpleEncryptor() {
+        val firstEncryptor = Encryptor()
+        val secondEncryptor = Encryptor()
+        updateEncryptors(firstEncryptor, secondEncryptor)
+        var message = "Simple message".toByteArray()
+        checkMessage(message, firstEncryptor, secondEncryptor)
+        message = "Русский текст".toByteArray()
+        checkMessage(message, firstEncryptor, secondEncryptor)
+    }
+
```
я бы на пустую строку тест добавил
кстати надо закостылить в ui это на всякий случай...

VKuzia commented on src/main/kotlin/by/dismess/core/utils/utils.kt:
```diff
@@ -12,3 +13,16 @@ fun hashMD5(input: String): BigInteger {
+
+fun byteToInt(number: Byte): Int {
+    return number.toInt()
+}
```
разве не проще на месте написать .toInt()

IgorFilipovich replied:
Проще)
Раньше просто было чуть по-другому

VKuzia commented on src/main/kotlin/by/dismess/core/security/SecureNetworkInterface.kt:
```diff
@@ -0,0 +1,72 @@
+        }
+        val encryptedMessage = sessionManager.encrypt(address, data)
+        networkInterface.sendRawMessage(address, encryptedMessage)
+    }
```
чот у нас интерфейс не интерфейсный...мб сделать таки интерфейс и таки InterfaceImpl?

VKuzia replied:
переименовать этот класс в Impl и интерфейс докинуть

IgorFilipovich replied:
Не, в этом и задумка, теперь мы будем инджектить не  NetworkInterface, а SecureNetworkInterface

VKuzia commented on src/main/kotlin/by/dismess/core/security/SecureNetworkInterface.kt:
```diff
@@ -0,0 +1,72 @@
+                    }
+                }
+            }
+        }
```
может, чтобы не было такой жёсткой вложенности, эту штуку сделать private fun и передавать в одну строчку?

VKuzia commented on src/main/kotlin/by/dismess/core/security/SessionManager.kt:
```diff
@@ -0,0 +1,56 @@
+import java.net.InetSocketAddress
+
+const val KEY_LIFETIME = 60000
+
```
я бы докинул _MS к имени этой переменной

VKuzia commented on src/main/kotlin/by/dismess/core/security/SessionManager.kt:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * Process message if it is a key or raw message, create new session if needed
+     * @return decrypted message, null if key was passed
```
мб Process message. If
а то я сломался в англ немножко

VKuzia commented on src/main/kotlin/by/dismess/core/security/SessionManager.kt:
```diff
@@ -0,0 +1,56 @@
+            addressToProtocolManager[address] = protocolManager
+        }
+        return protocolManager.processData(data)
+    }
```
мне немного не нравится, что тут функция как будто бы два поведения в себе сочетает...вместе с processData...типа и для ключа, и для сообщения. Может сделать функцию опознавания типа сообщения (просто иф из processData вынести и ифать при вызове, а не при обработке)? Так чуть читабельнее будет...хотя надо будет чуток методов наплодить. Что скажете?

IgorFilipovich replied:
Тогда нужно, чтобы SecureNetworkInterface знал про протокол, а это не оч хорошо

IgorFilipovich replied:
Возникла путаница с комментариями

VKuzia commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,142 @@
+    private fun cipherKey(aesOffset: Int, key: ByteArray): Key? {
+        val sessionNumber: Int = twoBytesToInt(key.sliceArray(0..2)) % (2 shl 16)
+        val sessionKey = findSession(sessionNumber) ?: return null
+        return generateCipherKey(sessionKey, aesOffset)
```
кажется, нигде потом эта нуловость не проверяется...может я ошибаюсь, но просто пустым всё становится. Разве это не должно означать, что что-то неправильно пошло? Или мы и хотим тогда передавать пустое сообщение? Кажется, тут есть место для каких-то ифов-ассертов

VKuzia commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,142 @@
+    private fun currentCipherKey(aesOffset: Int): Key? {
+        val sessionKey = findSession(currentSession) ?: return null
+        return generateCipherKey(sessionKey, aesOffset)
+    }
```
вот например эта штука вызовется и просто пустое сообщение будет....не уверен, что так и должно быть

VKuzia commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,142 @@
+        currentSession = twoBytesToInt(bytePublicKey.sliceArray(0..1))
+        bytePublicKey = bytePublicKey.sliceArray(2 until bytePublicKey.size)
+        val factory = KeyFactory.getInstance("DH")
+        val anotherKey = factory.generatePublic(X509EncodedKeySpec(bytePublicKey)) as DHPublicKey
```
можно в переменную вынести, как "AES"

IgorFilipovich committed:
SecureNetworkInterface PR fixes

Fix empty string problem
Style fixes

IgorFilipovich committed:
Fix SessionManager.processData comment

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,141 @@
+    private val cipherAlgo: String = "AES"
+    private var aesKeys: List<Pair<Int, ByteArray>> = mutableListOf()
+    private lateinit var publicKey: PublicKey
+    private var currentSession: Int = 0
```
Странно что перемешаны val и var. Стоит подумать где можно изменить var на val, val на const и переупорядочить.

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private fun cipherKey(aesOffset: Int, key: ByteArray): Key {
+        val sessionNumber: Int = twoBytesToInt(key.sliceArray(0..2)) % (2 shl 16)
```
twoBytesToInt использует 2 байта, а передается 3. Зачем брать по модулю? В sliceArray проблем нет, но зачем слайсы ради двух байт?) Обычный подмассив был бы эффективнее.

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,141 @@
+     */
+    private fun generateOffset(): Int {
+        return (0..MAX_OFFSET).random()
+    }
```
Уверен, что нужно включать верхнюю границу?

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,141 @@
+        if (aesKeys.size >= MAX_SESSIONS_SIZE) {
+            aesKeys = aesKeys.drop(1)
+        }
+        aesKeys = aesKeys + Pair<Int, ByteArray>(currentSession, newKey)
```
Зачем тут происходит копирование всего aesKeys? Если копирование не нужно, стоит использовать .add, если все-таки нужно, следует оставить комментарий.

Derovi commented on src/main/kotlin/by/dismess/core/security/Encryptor.kt:
```diff
@@ -0,0 +1,141 @@
+        val bytePublicKey = publicKey.encoded
+        var session = currentSession
+        if (updateSession) {
+            session += 1
```
*шутка про питон*

Derovi commented on src/main/kotlin/by/dismess/core/security/SecureNetworkInterface.kt:
```diff
@@ -0,0 +1,87 @@
+        try {
+            val encryptedMessage = sessionManager.encrypt(address, data)
+            networkInterface.sendRawMessage(address, encryptedMessage)
+        } catch (_: IllegalArgumentException) {
+        }
```
Странно выглядит пустой обработчик. Возможно, там стоит добавить printStackTrace. А если ошибка случается часто и это нормально, то просто скобки фигурные в одну строку)

Derovi commented on src/main/kotlin/by/dismess/core/utils/utils.kt:
```diff
@@ -12,3 +13,12 @@ fun hashMD5(input: String): BigInteger {
+
+fun intToBytes(number: Int, size: Int = 1): ByteArray {
+    return ByteBuffer.allocate(4).putInt(number).array().sliceArray((3 - size + 1)..3)
+}
```
В  котлине такие однострочные функции принято указывать через оператор = вместо { return ...}

Derovi commented on src/main/kotlin/by/dismess/core/utils/utils.kt:
```diff
@@ -12,3 +13,12 @@ fun hashMD5(input: String): BigInteger {
+}
+
+fun intToBytes(number: Int, size: Int = 1): ByteArray {
+    return ByteBuffer.allocate(4).putInt(number).array().sliceArray((3 - size + 1)..3)
```
3 + size - 1 = 2 + size. Если это было для понятности, то понятнее не стало))

Derovi commented on src/test/kotlin/by/dismess/core/security/SecureNetworkInterfaceTest.kt:
```diff
@@ -0,0 +1,74 @@
+            check(firstSecure, counter)
+            sleep(1L)
+        }
+    }
```
Не стоит делать такое в юнит тестах, они не должны прогоняться 10 секунд, надо искусственно как-то эмулировать тайм аут

IgorFilipovich committed:
Style fixes

IgorFilipovich committed:
Add possibility to set session lifetime Fix test not to wait for minute

IgorFilipovich committed:
Merge branch 'develop' of github.com:Derovi/dismess into secure-network-interface
